packages/@aws-cdk/aws-kinesisanalyticsv2/lib/flink-application.ts

```diff
@@ -0,0 +1,328 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as logs from '@aws-cdk/aws-logs';
+import * as core from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { ApplicationCode } from './application-code';
+import { CfnApplication, CfnApplicationCloudWatchLoggingOption } from './kinesisanalyticsv2.generated';
+import { flinkApplicationConfiguration, FlinkLogLevel, FlinkMetricsLevel } from './private/flink-application-configuration';
+import { validateFlinkApplicationProps } from './private/validation';
+import { PropertyGroup } from './property-group';
+
+export interface IFlinkApplication extends core.IResource, iam.IGrantable {
```

Comment:
I was worried that trying to share any code between a Flink application and an SQL application would encourage future CDK developers to look for similarities between the two resources. As it stands, it's bizarre that Flink applications and these SQL applications are even in the same CloudFormation module. The SQL application seems very integrated with Kinesis (e.g. it has a `KinesisStreamsInput` and `KinesisStreamsOutput`) but the Flink application doesn't really have any specific integration with Kinesis, nor would it necessarily use Kinesis. Mercifully, AWS keeps the Developer [Guide for Flink](https://docs.aws.amazon.com/kinesisanalytics/latest/java/what-is.html) and SQL applications separate. Trying to tease this hodgepodge apart seems like a great opportunity to add value with the CDK.

Thinking about what happens when someone adds SQLApplication to the CDK or when AWS adds a Samza application to Kinesis Analytics seems important. As it stands now, trying to use the Cfn documentation is rough users have to wade through a bunch of irrelevant properties. Adding another application type like Samza would add even more irrelevant properties when trying to setup Flink or SQL applications.

I wonder if I've done enough to keep FlinkApplication isolated. If more application types are added to this module a user will type `ka.` and see irrelevant properties when trying to setup a Flink application. Prefixing every type with `Flink` or `SQL` or `Samza` helps a bit which makes me think I should change my `PropertyGroups` type to `FlinkPropertyGroups`. Another thing to consider would be to get rid of the `aws-kinesisanalyticsv2` package and just replace it with `aws-flink-application` at the expense of clashing with the cloudformation namespaces. The readme for `aws-kinesisanalytics` could point users to `aws-flink-application` `aws-sql-application` (`aws-spark-application`?)and `aws-samza-application`.

At this point we only have 2 examples which makes this decision tough :/  I think if I had to choose right now I would prefer a different `aws-flink-application` package. Is there precedent for handling a situation like this in the CDK?

Reply:
I think there are 2 possible solutions here:

1. Have `IApplication`, `IFlinkApplication extends IApplication`, `FlinkApplication implements IFlinkApplication`, etc. inside the `aws-kinesisanalyticsv2` module. Then, we add new subinterfaces and classes inside the module as other application types get added.
2. Have a separate module per application type (so something like `aws-kinesisanalyticsv2-flink`, etc.). Perhaps a `aws-kinesisanalyticsv2` module too, which contains the common types (like `IApplication`).

I'll let you decide which one you prefer 🙂. Up to this point, we usually went with 1. in the CDK, but I think this case makes enough sense that we might want to go with 2.

The other advantage 2. has is the fact that they different modules can have different dependencies, which you can't have with 1.

Let me know what you think!

Reply:
I think your summary is spot on. After sleeping on it I'm leaning more and more to 2. My only follow up question is whether this should be `aws-kinesisanalyticsv2-flink` or whether we should just continue with the name `aws-kinesis-analytics-flink` without adding `aws-kinesisanalyticsv2`. 

If one or the other seems more inline with what's in CDK already I'd be happy to choose that. As it stands, if we can make these separate packages, perhaps there is no need to bother users with the v1/v2 distinction. I'll move forward without the v2 package for now that for now, but I'm still interested in your thoughts.

Reply:
I think you're right that probably at that point it's not worth it to bother customers with the V1/V2 distinction, and `aws-kinesis-analytics-flink` is most likely a better name for the module.

I love it!

Reply:
Ran with `aws-kinesisanalytics-flink` (no hyphen for `kinesis-analytics`) since this matched the naming convention of other modules like `aws-elasticloadbalancingv2-targets`.